Egress blog part 3

vadimeisenbergibm opened the pull request:
The third part of splitting #3179 into three parts. Continuation of #4196 and #4232.

vadimeisenbergibm commented:
/hold

vadimeisenbergibm commented on content/blog/2019/egress-traffic-control-in-istio-part-3/index.md:
```diff
@@ -0,0 +1,139 @@
+
+1.  Support for [TLS](https://en.wikipedia.org/wiki/Transport_Layer_Security) with
+    [SNI](https://en.wikipedia.org/wiki/Server_Name_Indication) or for [TLS origination](/docs/reference/glossary/#tls-origination) by Istio.
+1.  **Monitor** SNI and the source workload of every egress access.
```
@rcaballeromx Please note that I use the imperative in other bullets: support, monitor, prevent, enforce

vadimeisenbergibm commented on content/blog/2019/egress-traffic-control-in-istio-part-3/index.md:
```diff
@@ -0,0 +1,139 @@
+1.  **Prevent tampering**.
+1.  Traffic control is **transparent** to the applications.
+
+Now, let me present two alternative solutions for egress traffic control and show which requirements they satisfy, and,
```
@rcaballeromx I added "namely Kubernetes Network Policies and egress proxies and firewalls" in https://github.com/istio/istio.io/pull/4637/commits/3e7ef7662f8ceda57a1bec3a1340081bac2f7f0a

vadimeisenbergibm commented on content/blog/2019/egress-traffic-control-in-istio-part-3/index.md:
```diff
@@ -0,0 +1,139 @@
+Now, let me present two alternative solutions for egress traffic control and show which requirements they satisfy, and,
+more importantly, cannot satisfy.
+
+The most natural solution for egress traffic control is
```
@rcaballeromx Kubernetes Network Policies is the "native" Kubernetes solution. For a person familiar with Kubernetes, this would be the first mechanism to come to mind in the context of traffic control. However, it does not fulfill all the requirements. I am not trying to "sell Istio", and as part of my blog post I present the alternative solutions. This section is an answer to the questions I encountered when presenting egress traffic control in Istio: how is it better than Kubernetes Network Policies? why should I use Istio for egress traffic control and not a DNS-aware egress proxy?

vadimeisenbergibm replied:
Addressed by https://github.com/istio/istio.io/pull/4637/commits/1dbe203d93535b322fca096db0a2a5a20991ca3b.

vadimeisenbergibm commented on content/blog/2019/egress-traffic-control-in-istio-part-3/index.md:
```diff
@@ -0,0 +1,139 @@
+[Kubernetes Network Policies](https://kubernetes.io/docs/concepts/services-networking/network-policies/). Using
+Kubernetes Network Policies, cluster operators can specify which external services can be accessed by which pods. The
+pods may be identified by pod labels, namespace labels, or by IP ranges. The external services can be specified by IP
+ranges: Kubernetes Network Policies are not DNS-aware. The first requirement is satisfied since any TCP traffic can be
```
addressed by cd31534771739781c5b3bcae63f37454fd5d4ad9 and 3472526c5656c4ef4eeaf99b3e4d1ed4a421fd24

vadimeisenbergibm commented on content/blog/2019/egress-traffic-control-in-istio-part-3/index.md:
```diff
@@ -0,0 +1,139 @@
+pods may be identified by pod labels, namespace labels, or by IP ranges. The external services can be specified by IP
+ranges: Kubernetes Network Policies are not DNS-aware. The first requirement is satisfied since any TCP traffic can be
+controlled by Kubernetes Network Policies. The requirements 3 and 4 are satisfied partially: the policies can be
+specified per cluster or per pod, however the external services cannot be identified by domain names.
```
done in cf85e7b910519df608bd2cfcc5e1ba99091ec1b1

vadimeisenbergibm commented on content/blog/2019/egress-traffic-control-in-istio-part-3/index.md:
```diff
@@ -0,0 +1,139 @@
+Kubernetes Network Policies, cluster operators can specify which external services can be accessed by which pods. The
+pods may be identified by pod labels, namespace labels, or by IP ranges. The external services can be specified by IP
+ranges: Kubernetes Network Policies are not DNS-aware. The first requirement is satisfied since any TCP traffic can be
+controlled by Kubernetes Network Policies. The requirements 3 and 4 are satisfied partially: the policies can be
```
done in cf85e7b

vadimeisenbergibm commented on content/blog/2019/egress-traffic-control-in-istio-part-3/index.md:
```diff
@@ -0,0 +1,139 @@
+approach, applications are configured to direct the traffic to the proxy and to use some proxy protocol, e.g.
+[SOCKS](https://en.wikipedia.org/wiki/SOCKS).
+Since the applications must be specially configured, this solution is not transparent. Moreover, egress proxies are not
+Kubernetes-aware, since neither pod labels nor pod service account are known to the egress proxy. Such egress proxies
```
@rcaballeromx Good idea to use operators as the subjects. Good that you did not use **cluster** operators, since the egress proxies are not a Kubernetes concept, so Kubernetes cluster operators not necessarily configure them, some other operators do.

vadimeisenbergibm commented:
@rcaballeromx Thank you very much for your review, very helpful as usual. I applied your comments and made some changes to the remaining sections.

I will strive to avoid using passive voice at all cost. In this blog I left: "is based on", "is integrated with", "you are convinced", "with Istio sidecars injected",  since I think these are valid uses. I use "you" multiple times to address the reader directly.

I have a summary table with the features and the alternatives, so I did not add "In summary" sentence before the summary table.

rcaballeromx commented on content/blog/2019/egress-traffic-control-in-istio-part-3/index.md:
```diff
@@ -0,0 +1,136 @@
+security of your cluster.
+I also hope that I managed to convince you that Istio can serve as an effective tool for controlling egress traffic
+securely, and that Istio has multiple advantages over the alternative solutions.
+In my opinion, you can even choose secure control of egress traffic as the first use case for applying Istio to your
```
```suggestion
In my opinion, secure control of egress traffic is a great choice if you are looking for your first Istio use case.
```

rcaballeromx commented on content/blog/2019/egress-traffic-control-in-istio-part-3/index.md:
```diff
@@ -0,0 +1,136 @@
+I also hope that I managed to convince you that Istio can serve as an effective tool for controlling egress traffic
+securely, and that Istio has multiple advantages over the alternative solutions.
+In my opinion, you can even choose secure control of egress traffic as the first use case for applying Istio to your
+cluster.
```
```suggestion

```

rcaballeromx commented on content/blog/2019/egress-traffic-control-in-istio-part-3/index.md:
```diff
@@ -0,0 +1,136 @@
+securely, and that Istio has multiple advantages over the alternative solutions.
+In my opinion, you can even choose secure control of egress traffic as the first use case for applying Istio to your
+cluster.
+Istio will already be beneficial for you, even before you start using all other features, such as
```
```suggestion
In this case, Istio already provides you some benefits even before you start using all other Istio features:
```

vadimeisenbergibm replied:
@rcaballeromx some benefits does not sound right to me. Istio provides a single benefit of secure control of egress traffic, if it is the first use case.

vadimeisenbergibm replied:
@rcaballeromx I have applied "some benefits", though I have doubts about it, see https://github.com/istio/istio.io/pull/4637#discussion_r303580907

rcaballeromx commented on content/blog/2019/egress-traffic-control-in-istio-part-3/index.md:
```diff
@@ -0,0 +1,136 @@
+In my opinion, you can even choose secure control of egress traffic as the first use case for applying Istio to your
+cluster.
+Istio will already be beneficial for you, even before you start using all other features, such as
+traffic management, security, policies and telemetry, applied to traffic between microservices inside the cluster.
```
Please add links to the concepts for the features you mention here.

vadimeisenbergibm replied:
Done in https://github.com/istio/istio.io/pull/4637/commits/0dbe862cb8867e1e23cc3e231a4a76be9c3df0b4

rcaballeromx commented on content/blog/2019/egress-traffic-control-in-istio-part-3/index.md:
```diff
@@ -0,0 +1,136 @@
+cluster.
+Istio will already be beneficial for you, even before you start using all other features, such as
+traffic management, security, policies and telemetry, applied to traffic between microservices inside the cluster.
+You should pay attention, however, to performance overhead of Istio egress traffic control and measure it for your use
```
I would remove this sentence. It ends the post on a negative note. You already mentioned the performance considerations and I think it would be better to end on the positive note of all the features Istio provides to readers and how secure control of egress traffic can serve as a starting point.

vadimeisenbergibm replied:
Done in https://github.com/istio/istio.io/pull/4637/commits/3c3e8aa3fb3dc258470c6eda7aadcb99eebe22ad

vadimeisenbergibm commented on content/blog/2019/egress-traffic-control-in-istio-part-3/index.md:
```diff
@@ -0,0 +1,136 @@
+Istio egress traffic control is **transparent** with regard to TLS traffic, since Istio is transparent:
+you do not need to change the applications or to configure their containers.
+For HTTP traffic with TLS origination, you must configure the applications to use HTTP instead of HTTPS
+when the applications run with Istio sidecars injected.
```
I mean "the applications that run as part of Istio service mesh". Maybe it is clear from the context, the stress here is that once you run an application as part of a service mesh (Istio sidecar is injected, manually or automatically), you have to configure that application explicitly to use HTTP instead of HTTPS.

vadimeisenbergibm replied:
addressed in https://github.com/istio/istio.io/pull/4637/commits/8f7d21ef7e3300a9f4c9a4cc8c885dd6755dbbd9.

vadimeisenbergibm commented on content/blog/2019/egress-traffic-control-in-istio-part-3/index.md:
```diff
@@ -0,0 +1,136 @@
+Istio egress traffic control is **secure**: it is based on the strong identity of Istio and, when you
+apply
+[additional security measures](/docs/tasks/traffic-management/egress/egress-gateway/#additional-security-considerations),
+it is tamper-proof.
```
right, resilient to tampering is less pretentios.

vadimeisenbergibm commented on content/blog/2019/egress-traffic-control-in-istio-part-3/index.md:
```diff
@@ -0,0 +1,136 @@
+   need to learn a single policy and configuration language for all types of traffic.
+*  Istio egress traffic control is integrated with Istio policy and telemetry adapters and can work out-of-the-box.
+*  When you use external monitoring/access control systems with Istio, you must write the adapters for them only once,
+   and then apply the adapters for all types of traffic: ingress, egress and in-cluster.
```
@rcaballeromx I have doubts regarding "the adapters are applied **to** traffic". The adapters are applied with regard to traffic, or for traffic. For example, apply an adapter to control egress traffic or to perform monitoring of egress traffic.

vadimeisenbergibm replied:
@rcaballeromx It is similar to your suggestion: 
> Use Istio's [traffic management features](/docs/concepts/traffic-management/) for egress traffic:

So you use features **for** egress traffic, apply adapters **for** egress traffic.

vadimeisenbergibm replied:
addressed by https://github.com/istio/istio.io/pull/4637/commits/66d5805078d550e156788c160cd81ecbbb433685 and https://github.com/istio/istio.io/pull/4637/commits/c9366c846077de11b1c3405bc9a6e90a9569161c

vadimeisenbergibm commented on content/blog/2019/egress-traffic-control-in-istio-part-3/index.md:
```diff
@@ -0,0 +1,136 @@
+
+Let me summarize the features of Istio egress traffic control and of the alternative solutions in the following table:
+
+| | Kubernetes Network Policies | Legacy Egress Proxy or Firewall | Istio Egress Traffic Control |
```
Done in https://github.com/istio/istio.io/pull/4637/commits/02e8e322a88213a7719a14e7ccec97e2471bdb02

vadimeisenbergibm commented on content/blog/2019/egress-traffic-control-in-istio-part-3/index.md:
```diff
@@ -0,0 +1,136 @@
+
+| | Kubernetes Network Policies | Legacy Egress Proxy or Firewall | Istio Egress Traffic Control |
+| --- | --- | --- | ---|
+| DNS-aware | {{< cancel_icon >}} | {{< checkmark_icon >}} | {{< checkmark_icon >}} |
```
Done in https://github.com/istio/istio.io/pull/4637/commits/02e8e322a88213a7719a14e7ccec97e2471bdb02

rcaballeromx commented on content/blog/2019/egress-traffic-control-in-istio-part-3/index.md:
```diff
@@ -70,67 +69,72 @@ are not transparent and not Kubernetes-aware.
+In my opinion, secure control of egress traffic is a great choice if you are looking for your first Istio use case.
+In this case, Istio already provides you some benefits even before you start using all other Istio features:
+[traffic management](/docs/tasks/traffic-management/), [security](/docs/tasks/security/),
+[policies](/docs/tasks/policy-enforcement/) and [telemetry](/docs/tasks/telemetry/), applied to traffic between microservices inside the cluster.
```
@vadimeisenbergibm What is the final message you want to give readers after they have followed you through all three parts? The message on this finishing paragraph is a bit muddled. Try to picture someone who followed the entire journey. They learned what controlling egress traffic securely was all about, the different attacks that make it important, and the different ways they can protect against said attacks. After all of that, you reached the conclusion that Istio's secure control of egress traffic is the way to go and explained why. 
IMO, this paragraph should communicate three things to your readers:
1. What's next? Assume readers have followed along and secured their egress traffic, what specific actions would you recommend they take next?
2. Where can I learn more? I think this is what you were trying to do by listing the features. However, those still feel disconnected from the topic at hand. What tasks do you recommend they complete next? What reference materials will allow your readers to do more around traffic management?
3. Summarize the value that Istio gives them and that they cannot get any other way.
Some of this things are alluded to so you might only need to rearrange and reword a little.

vadimeisenbergibm replied:
Addressed in https://github.com/istio/istio.io/pull/4637/commits/f2a4c183b96d82deed21f3576e46b220063c9deb and https://github.com/istio/istio.io/pull/4637/commits/667c470c82b0a25c882b12c5b40d02b11647a7b6

vadimeisenbergibm commented on content/blog/2019/egress-traffic-control-in-istio-part-3/index.md:
```diff
@@ -70,67 +69,72 @@ are not transparent and not Kubernetes-aware.
-traffic control. I would encourage you to measure carefully different configurations for your own applications and your
-external services, and decide whether you can afford the performance overhead for your use cases. You should weigh the
-required level of security versus your performance requirements, and also compare with the performance overhead of
+adding a third proxy between the application and the external service. Since the traffic between the egress gateway's proxy and
```
Done in https://github.com/istio/istio.io/pull/4637/commits/e7924cced9b20141bbf6a3736028099e4685b4e8

rcaballeromx commented on content/blog/2019/egress-traffic-control-in-istio-part-3/index.md:
```diff
@@ -0,0 +1,147 @@
+security of your cluster.
+Hopefully, I also managed to convince you that Istio is an effective tool to control egress traffic
+securely, and that Istio has multiple advantages over the alternative solutions.
+If you need to control egress traffic in a secure and transparent way, specifying external services as domain names and
```
```suggestion
Istio is the only solution I'm aware of that lets you:

- Control egress traffic in a secure and transparent way
- Specify external services as domain names

```

rcaballeromx commented on content/blog/2019/egress-traffic-control-in-istio-part-3/index.md:
```diff
@@ -0,0 +1,147 @@
+Hopefully, I also managed to convince you that Istio is an effective tool to control egress traffic
+securely, and that Istio has multiple advantages over the alternative solutions.
+If you need to control egress traffic in a secure and transparent way, specifying external services as domain names and
+specifying traffic source by Kubernetes artifacts, Istio is the only solution for you I am aware of.
```
```suggestion
- Use Kubernetes artifacts to specify a traffic source.
```

rcaballeromx commented on content/blog/2019/egress-traffic-control-in-istio-part-3/index.md:
```diff
@@ -0,0 +1,147 @@
+[policies](/docs/tasks/policy-enforcement/) and [telemetry](/docs/tasks/telemetry/), applied to traffic between
+microservices inside the cluster.
+
+So, if you did not yet have a chance to work with Istio, [install](/docs/setup/kubernetes/install/) it in your cluster
```
```suggestion
So, if you haven't had the chance to work with Istio yet, [install Istio](/docs/setup/kubernetes/install/) on your cluster
```

rcaballeromx commented on content/blog/2019/egress-traffic-control-in-istio-part-3/index.md:
```diff
@@ -0,0 +1,147 @@
+microservices inside the cluster.
+
+So, if you did not yet have a chance to work with Istio, [install](/docs/setup/kubernetes/install/) it in your cluster
+and check [egress traffic control](/docs/tasks/traffic-management/egress/) and other
```
```suggestion
and check our [egress traffic control task](/docs/tasks/traffic-management/egress/) and the tasks for the other
```

rcaballeromx commented on content/blog/2019/egress-traffic-control-in-istio-part-3/index.md:
```diff
@@ -0,0 +1,147 @@
+
+So, if you did not yet have a chance to work with Istio, [install](/docs/setup/kubernetes/install/) it in your cluster
+and check [egress traffic control](/docs/tasks/traffic-management/egress/) and other
+[Istio features](https://istio.io/docs/tasks/). Tell us what you think at [discuss.istio.io](https://discuss.istio.io).
```
```suggestion
[Istio features](https://istio.io/docs/tasks/). We also want to hear from you, please join us at [discuss.istio.io](https://discuss.istio.io).
```

vadimeisenbergibm commented:
@rcaballeromx Thank you for your proposals, I applied all of them, and added three more commits.

specify a traffic source -> specify the traffic source, since traffic in our cases has a single source and a single destination  
https://github.com/istio/istio.io/pull/4637/commits/df242b28f1bacb6712e9cce2ee7b0a3f69ba507c

and
egress task -> egress tasks, there are multiple tasks at the link
https://github.com/istio/istio.io/pull/4637/commits/064a3fc30be2400685a524deac1320118b28d59b

and removed the final dot from the last bullet: https://github.com/istio/istio.io/pull/4637/commits/3811e766411a3ee417b75c0e8813f0c965bfb78e
